streams/src/main/java/org/apache/kafka/streams/state/internals/buffer/InMemoryTimeOrderedKeyValueBufferStore.java

```diff
@@ -0,0 +1,219 @@
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals.buffer;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.BytesSerializer;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+import org.apache.kafka.streams.processor.internals.ProcessorStateManager;
+import org.apache.kafka.streams.processor.internals.RecordBatchingStateRestoreCallback;
+import org.apache.kafka.streams.processor.internals.RecordCollector;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.apache.kafka.streams.state.internals.ContextualRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.Map;
+
+public class InMemoryTimeOrderedKeyValueBufferStore implements StateStore {
+    private static final Logger LOG = LoggerFactory.getLogger(InMemoryTimeOrderedKeyValueBufferStore.class);
+    private static final String BUFFER_TIME_HEADER = "_bufferTime";
+    private static final BytesSerializer KEY_SERIALIZER = new BytesSerializer();
+    private static final ByteArraySerializer VALUE_SERIALIZER = new ByteArraySerializer();
+
+    private final InMemoryTimeOrderedKeyValueBuffer buffer;
+    private final String storeName;
+
+    private RecordCollector collector;
+    private String changelogTopic;
+    private boolean open;
+
+    public static class Builder implements StoreBuilder<StateStore> {
+        private final String storeName;
+
+        public Builder(final String storeName) {
+            this.storeName = storeName;
+        }
+
+        @Override
+        public StoreBuilder<StateStore> withCachingEnabled() {
+            return null;
```

Comment:
some of these seemed best to just ignore.

Reply:
All methods returning `null` seems to be invalid to call -- might be better to `throw new UnsupportedOperationException()`?

Note, that this part of the interface was designed to let users create store manually (for Processor API usage). Not sure, if we want to allow users to plug in custom stores for the suppress buffer at some point. But this might be follow up work. Just mention it for completeness. We would need to add a public "BufferStoreInterface" for this and would need another KIP.

Reply:
Yes, I intentionally did not propose to create this API right now, to give us a release cycle to live with the API privately, and maybe change our minds about how it should look.

Ultimately, I think it will have to be configurable.

Reply:
I opted to just `return this;` for the ignored settings instead of throwing.

Reply:
It's an internal API atm, right? Thus, if our code does not call it, nobody calls it? Hence, throwing would allow us to detect bugs.

Reply:
ok